elm/diff_model.py

```diff
@@ -70,12 +75,153 @@ def unsafe_execute(code_str: str, timeout: int = 5):
+    func_name: str  # the name of the function that we want to execute
+    import_line: str  # the import lines we add to the code
+    func_preamble: str  # the function definition plus possibly a few initial lines to generate codes
+    return_line: str  # the return line we add to the end of the code
+
+    def __init__(self, cfg, sandbox_server='http://localhost:5000') -> None:
+        if isinstance(cfg, str):
+            self.cfg = OmegaConf.load(cfg)
+        elif isinstance(cfg, (dict, DictConfig)):
+            self.cfg = DictConfig(cfg)
+        else:
+            raise ValueError
+
         set_seed(self.cfg.seed)
+        # Use RNG to rotate random seeds during inference.
+        self.rng = np.random.default_rng(seed=self.cfg.seed)
+        self.sandbox_server = sandbox_server
         self.model, self.tokenizer = model_setup(self.cfg)
 
+    def generate_prompt_str(self, seed: str, tokenizer=None, batch_size=None,
+                            append_return=True, without_trunc=False) -> list[str]:
+        """
+        Args:
+            seed: the seed text.
+            tokenizer: (Optional) assign only if you want to use a different tokenizer (default: None)
+            batch_size: (Optional) override the batch size in config.
+            append_return: (Optional) append a return line to the code in the end.
+            without_trunc: (Optional) True if we don't apply the `truncate` function.
+        Returns:
+            a list of code(s) generated by the model.
+        """
+        tokenizer = self.tokenizer if tokenizer is None else tokenizer
+        encoding = tokenizer(
+            [seed + '\n\n' + self.func_preamble],
+            truncation=True,
+            padding=True,
+            max_length=self.cfg.gen_max_len,
+            return_tensors="pt",
+        )
+
+        cfg = OmegaConf.merge(self.cfg, {"batch_size": self.cfg.batch_size if batch_size is None else batch_size})
+        with torch.no_grad():
+            completion = sample(cfg, self.model, self.tokenizer, encoding)
+        # Reset random seed
+        set_seed(int(self.rng.integers(0, 1e8)))
+
+        if without_trunc:
+            truncation = completion
+        else:
+            truncation = [truncate(code, print_num=float('inf'), only_local_scope=True) for code in completion]
+
+        truncation = [self.import_line + '\n' + self.func_preamble + '\n' + code for code in truncation]
+
+        if append_return:
+            truncation = [code + '\n' + self.return_line for code in truncation]
+
+        return truncation
+
+    def generate_program(self, code: str) -> list[dict]:
+        """
+        Given a piece of code, do prompt mutation, call the sandbox server to execute the code and return the result.
+        Args:
+            code: the full code string.
+        Returns:
+            a numpy array (if successful) or the exception object.
+        """
+        results = []
+        for code in self.generate_prompt_str(code):
```

Comment:
You can send a batch of code strings to the sandbox, which would probably faster than making many requests

Reply:
Yeah that's a great point! I saw your sandbox supports batches now!

Reply:
Yup batches although no parallelism yet :)

Reply:
@TheExGenesis 
Question for your batch return: what happens if you have a mixture of good and bad sodaracers? From your code you will collect the dict in the first place and always put a 200 as the return code? (also can `bad_request(...)` be unraveled like a tuple?)
For the image task batch return was also not implemented. I start to feel that it's gonna take some effort....

Reply:
We could totally refactor the error code into your dict, but there will be a small little bit of interfacing on both of our sides.
Maybe we leave it for another PR?